lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,768 @@
+          {:shutdown, state} -> {:stop, :shutdown, state}
+        end
+
+      # We may hit clause if we send $gen_restart and then
+      # someone calls terminate_child, removing the child.
+      %{} ->
+        {:noreply, state}
+    end
+  end
+
+  def handle_info(msg, state) do
+    :error_logger.error_msg('DynamicSupervisor received unexpected message: ~p~n', [msg])
+    {:noreply, state}
+  end
+
+  @impl true
+  def code_change(_, %{mod: mod, args: args} = state, _) do
+    case mod.init(args) do
+      {:ok, flags} when is_map(flags) ->
+        case init(state, flags) do
+          {:ok, state} -> {:ok, state}
+          {:error, reason} -> {:error, {:supervisor_data, reason}}
+        end
+
+      :ignore ->
+        {:ok, state}
+
+      error ->
+        error
+    end
+  end
+
+  @impl true
+  def terminate(_, %{children: children} = state) do
+    :ok = terminate_children(children, state)
+  end
+
+  defp terminate_children(children, state) do
+    {pids, times, stacks} = monitor_children(children)
+    size = map_size(pids)
+
+    timers =
+      Enum.reduce(times, %{}, fn {time, pids}, acc ->
+        Map.put(acc, :erlang.start_timer(time, self(), :kill), pids)
+      end)
+
+    stacks = wait_children(pids, size, timers, stacks)
+
+    for {pid, {child, reason}} <- stacks do
+      report_error(:shutdown_error, reason, pid, child, state)
+    end
+
+    :ok
+  end
+
+  defp monitor_children(children) do
+    Enum.reduce(children, {%{}, %{}, %{}}, fn
+      {_, {:restarting, _}}, acc ->
+        acc
+
+      {pid, {_, restart, _, _, _} = child}, {pids, times, stacks} ->
+        case monitor_child(pid) do
+          :ok ->
+            times = exit_child(pid, child, times)
+            {Map.put(pids, pid, child), times, stacks}
+
+          {:error, :normal} when restart != :permanent ->
+            {pids, times, stacks}
+
+          {:error, reason} ->
+            {pids, times, Map.put(stacks, pid, {child, reason})}
+        end
+    end)
+  end
+
+  defp monitor_child(pid) do
+    ref = Process.monitor(pid)
+    Process.unlink(pid)
+
+    receive do
+      {:EXIT, ^pid, reason} ->
+        receive do
+          {:DOWN, ^ref, :process, ^pid, _} -> {:error, reason}
+        end
+    after
+      0 -> :ok
+    end
+  end
+
+  defp exit_child(pid, {_, _, shutdown, _, _}, times) do
+    case shutdown do
+      :brutal_kill ->
+        Process.exit(pid, :kill)
+        times
+
+      :infinity ->
+        Process.exit(pid, :shutdown)
+        times
+
+      time ->
+        Process.exit(pid, :shutdown)
+        Map.update(times, time, [pid], &[pid | &1])
+    end
+  end
+
+  defp wait_children(_pids, 0, timers, stacks) do
+    for {timer, _} <- timers do
+      _ = :erlang.cancel_timer(timer)
+
+      receive do
+        {:timeout, ^timer, :kill} -> :ok
+      after
+        0 -> :ok
+      end
+    end
+
+    stacks
+  end
+
+  defp wait_children(pids, size, timers, stacks) do
+    receive do
```

Comment:
Do we want to discard other messages to avoid growing message queue?

Reply:
It is tricky because we need to make sure to not read any call or system message.

Reply:
But we are only calling this in terminate/2 so we can throw those away safely - unless we want the messages in the proc_lib crash report but there should never be one for a supervisor. A DynamicSupervisor is a prime target for a lot of calls and could be slow to shutdown, everytime we do a selective receive we will have to read these messages. Its ok to leave this.

Reply:
Same as above, this code is called by terminate_child.